server/src/main/java/org/opensearch/tasks/TaskResourceTracker.java

```diff
@@ -0,0 +1,153 @@
+                true, threadpoolName);
+
+        taskMap.get(new TaskInfoKey(taskId)).add(taskWorkerResourceUtilInfo);
+    }
+
+    public void unregisterWorkerForTask(long taskId, long workerId, long cpuCurrent, long bytesCurrent) {
+        // This happens when task is unregistered first and then the runnable finishes
+        if (taskMap.get(new TaskInfoKey(taskId)) == null) {
+            return;
+        }
+        TaskWorkerResourceUtilInfo taskWorkerResourceUtilInfo = taskMap.get(new TaskInfoKey(taskId)).stream()
+            .filter(twrui -> twrui.getWorkerId() != null && twrui.getWorkerId() == workerId
+                && twrui.isActive()).findFirst().get();
+        taskWorkerResourceUtilInfo.setHeapNow(bytesCurrent);
+        taskWorkerResourceUtilInfo.setCpuNow(cpuCurrent);
+        taskWorkerResourceUtilInfo.setActive(false);
+    }
+
+    private void snapshot() {
+        final long[] search = {0};
+        final long[] response = {0};
+        taskMap.forEach((taskInfoKey, taskWorkerResourceUtilInfos) -> taskWorkerResourceUtilInfos.forEach(taskWorkerResourceUtilInfo -> {
+            if (taskWorkerResourceUtilInfo.isActive()) {
+                taskWorkerResourceUtilInfo.setHeapNow(threadMXBean.getThreadAllocatedBytes(taskWorkerResourceUtilInfo.getWorkerId()));
+            }
+            if ("tw".equals(taskWorkerResourceUtilInfo.getThreadPoolName())) {
+                response[0] += taskWorkerResourceUtilInfo.getOverheardBytes();
+            } else {
+                search[0] += taskWorkerResourceUtilInfo.getHeapNow() - taskWorkerResourceUtilInfo.getHeapStart();
+            }
+        }));
+
+        searchBytes = search[0];
+        responseOverheadBytes = response[0];
+    }
+
+    @Override
+    public void init() throws IOException {
+    }
+
+    @Override
+    public void checkAndNotify() throws IOException {
+        logger.trace("Taking Scheduled snapshot for System resources");
+        snapshot();
+    }
+
+    public void unregisterTaskForTracking(long id) {
+        logger.trace("Unregistering task");
+        // TODO if there's any active worker take snapshot of those and exit. (If accurate data required for a task)
+        taskMap.remove(new TaskInfoKey(id));
+    }
+
+    public void registerResponseOverhead(long taskId, long bytes) {
+        TaskWorkerResourceUtilInfo taskWorkerResourceUtilInfo =
+            new TaskWorkerResourceUtilInfo(bytes);
+
+        // TODO remove from overhead map as well
+
+        taskMap.get(new TaskInfoKey(taskId)).add(taskWorkerResourceUtilInfo);
+    }
+
+    public void registerResponseOverhead1(Object ob, long bytes) {
+        overhead.put(ob, bytes);
+    }
+
+    public void transfer(long taskId, Object ob, long bytes) {
+        TaskInfoKey key = new TaskInfoKey(taskId);
+        if (!overhead.containsKey(ob) || taskMap.get(key) == null) return;
+
+        long bytesStart = overhead.get(ob);
```

Comment:
NPE here, if the key gets removed in between check and usage. Use `Long` instead